lib/handler/connect.c

```diff
@@ -238,65 +294,182 @@ static void on_connect_timeout(h2o_timer_t *entry)
     on_connect_error(self, h2o_httpclient_error_io_timeout);
 }
 
-static void store_server_addresses(struct st_connect_generator_t *self, struct addrinfo *res)
+static void set_last_error(struct st_connect_generator_t *self, enum error_class class, const char *str)
 {
+    if (self->last_error.class <= class) {
+        self->last_error.class = class;
+        self->last_error.str = str;
+    }
+}
+
+static void on_resolution_delay_timeout(h2o_timer_t *entry)
+{
+    struct st_connect_generator_t *self = H2O_STRUCT_FROM_MEMBER(struct st_connect_generator_t, eyeball_delay, entry);
+
+    assert(self->server_addresses.used == 0);
+
+    try_connect(self);
+}
+
+static void on_connection_attempt_delay_timeout(h2o_timer_t *entry)
+{
+    struct st_connect_generator_t *self = H2O_STRUCT_FROM_MEMBER(struct st_connect_generator_t, eyeball_delay, entry);
+
+    /* If no more address is available, continue trying the current attempt until the connect_timeout expires. */
+    if (self->server_addresses.used == self->server_addresses.size)
+        return;
+
+    /* close current connection attempt and try next. */
+    h2o_socket_close(self->sock);
+    self->sock = NULL;
+    try_connect(self);
+}
+
+static int store_server_addresses(struct st_connect_generator_t *self, struct addrinfo *res)
+{
+    int added = 0;
+
     /* copy first entries in the response; ordering of addresses being returned by `getaddrinfo` is respected, as ordinary clients
      * (incl. forward proxy) are not expected to distribute the load among the addresses being returned. */
-    do {
-        struct st_server_address_t *dst = self->server_addresses.list + self->server_addresses.size++;
-        dst->sa = h2o_mem_alloc_pool_aligned(&self->src_req->pool, H2O_ALIGNOF(struct sockaddr), res->ai_addrlen);
-        memcpy(dst->sa, res->ai_addr, res->ai_addrlen);
-        dst->salen = res->ai_addrlen;
-    } while (self->server_addresses.size < PTLS_ELEMENTSOF(self->server_addresses.list) && (res = res->ai_next) != NULL);
+    while (self->server_addresses.size < PTLS_ELEMENTSOF(self->server_addresses.list) && res != NULL) {
+        if (h2o_connect_lookup_acl(self->handler->acl.entries, self->handler->acl.count, res->ai_addr)) {
+            struct st_server_address_t *dst = self->server_addresses.list + self->server_addresses.size++;
+            dst->sa = h2o_mem_alloc_pool_aligned(&self->src_req->pool, H2O_ALIGNOF(struct sockaddr), res->ai_addrlen);
+            memcpy(dst->sa, res->ai_addr, res->ai_addrlen);
+            dst->salen = res->ai_addrlen;
+            added = 1;
+        }
+        res = res->ai_next;
+    }
+
+    return added;
```

Comment:
`added == 0` can also happen if the server list has no room to store new addresses, leading to an incorrect classification of `ERROR_CLASS_ACCESS_PROHIBITED`

Conversely, multiple addresses could be delivered by the `_getaddr` call but only a subset might have been prohibited by acl and would be missed by the `if (!store_server_addresses(self, res))` check.

Reply:
I do not think so.

`store_server_addresses` is only called when `getaddrinfo` returns successfully, in which case there would be at least one address being obtained.

Quoting [RFC 3493](https://datatracker.ietf.org/doc/html/rfc3493.html) section 6.1:
> A zero return value for getaddrinfo() indicates successful completion; a non-zero return value indicates failure.  The possible values for the failures are listed below under Error Return Values.
> 
> Upon successful return of getaddrinfo(), the location to which res points shall refer to a linked list of addrinfo structures, each of which shall specify a socket address and information for use in creating a socket with which to use that socket address.  The list shall include at least one addrinfo structure.

When address lookup returns no address, `errstr` will be non-NULL, and the error being returned will be recorded and reported.

At least that's the intent of the code.

I think the "nxdomain" subtest in t/50connect-proxy-status.t covers such case?

Reply:
Consider a case where the v4 getaddr call returns many entries, (eg. MAX_CONNECT_RETRIES or more), then v6 call returns with a few addresses, `store_server_addresses()` would not enter the `while (self->server_addresses.size < PTLS_ELEMENTSOF(self->server_addresses.list) && res != NULL) ` loop and `added == 0`

I had bumped MAX_CONNECT_RETRIES to `24` because I ran into a few examples where the v4 response entries would fill the list and no v6 addresses would be stored. I'm not actually sure what the max entries returned from `getaddrinfo()` are.

My second case is invalid because if at least one entry did not match the acl then the response should not be prohibited.

Reply:
This was one i ran into
```
> dig +short a quantcount.com | wc  -l
56
```

Reply:
Thank you for the clarification. I clearly missed your first point which indeed is a valid argument.

To address the issue, in c8a123d, I've reduced the number of maximum entries to 8, and limited the number of addresses that can be retained for each address family to 4. With this change, there would be room for the lookup that complete later, enough that v4 and v6 addresses can be tried one by one.

If we think that 4 * 2 might not be enough in some cases, we can make that a configurable knob. WDYT?

Regarding the maximum number of addresses that can be returned by DNS, my understanding is that a conservative answer is somewhere below 30 (i.e., 512-byte UDP payload, minus some overhead, divided by 16 (which is the space required to store one IPv4 address)). But as said, I do not think we need to try all addresses. It is hard to expect many of the return addresses to be unreachable. Trying out too many might also be considered as an amplification vector.